dashboard/utils.py

```diff
@@ -241,25 +242,66 @@ def get_payments_count_for_course(self, course):
             course_key__in=course.courserun_set.values('edx_course_key'),
         ).values('order_id').distinct().count()
 
-    def get_custom_number_of_attempts_for_course(self, course):
+    def get_number_of_attempts_left(self, course):
         """
-        Get a custom number of exam attempts for given course. If payment was made before
-        the first date, learner get 2 attempts. Otherwise, 1.
+        Checks for each payment from before first date (when a payment provided two attempts)
+        and find if any unused attempts are carried over to the single attempt era (when a payment
+        provides one attempt).
+        Calculate the number of attempts remaining based on the calculated carry over and the other
+        payments and attempts.
+
         Args:
             course (courses.models.Course): a course
         Returns:
-            int: a number of attempts
+            int: a number of attempts left
         """
-        lines = Line.objects.filter(
+        first_date = course.program.exam_attempts_first_date
+        payments_qset = Line.objects.filter(
             order__status__in=Order.FULFILLED_STATUSES,
             order__user=self.user,
-            course_key__in=course.courserun_set.values('edx_course_key'),
+            course_key__in=course.courserun_set.values('edx_course_key')
         ).order_by('created_at')
-        first_date = course.program.exam_attempts_first_date
-        num_attempts = sum([ATTEMPTS_PER_PAID_RUN_OLD
-                            if line.modified_at < first_date else ATTEMPTS_PER_PAID_RUN for line in lines])
+        used_attempts_qset = ExamAuthorization.objects.filter(user=self.user, course=course, exam_taken=True)
+        # if for some reason the first date is not set, return the difference between payments and used attempts
+        if first_date is None:
+            return payments_qset.count() - used_attempts_qset.count()
+
+        # number of payments before the fist date
+        old_payments = payments_qset.filter(modified_at__lt=first_date).count()
+        # number of payments after the first date
+        new_lines = payments_qset.filter(modified_at__gte=first_date).count()
+
+        # number of used attempts before the second date
+        old_attempts = used_attempts_qset.filter(
+            exam_run__date_first_eligible__lt=self.program.exam_attempts_second_date
```

Comment:
there is a check for first_date above, and there is an assumption that both of them have to be set, because that is how the transition was made to single attempt per payment.

Reply:
I see, If this was assumed in the previous work it's fine. Although we should have not allowed blank/null for this date making sure that Django admin doesn't allow empty end dates. I'm fine either way. But I'm fine either way. Thanks for clarifying that.

Reply:
it would be null for programs that don't have any exams